### DISCUSSION TITLE
Lesson 7: Changes not staged for commit [lib/chainlink-brownie-contracts ] ?

### DISCUSSION THREAD

**Discussion by LareineHan** [Cyfrin/foundry-full-course-cu#870] [Q&A] [answered]
##lib/chainlink-brownie-contracts 

It was abnormally shown when I command 
```zsh
git status
```
```zsh
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   .gas-snapshot
        modified:   .gitignore
        modified:   .gitmodules
        new file:   Makefile
        modified:   README.md
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694712657.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694712664.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694712707.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694713369.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694713382.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-1694713404.json
        new file:   broadcast/DeployFundMe.s.sol/11155111/run-latest.json
        modified:   foundry.toml
        new file:   lib/chainlink-brownie-contracts
        new file:   lib/foundry-devops
        deleted:    script/Counter.s.sol
        new file:   script/DeployFundMe.s.sol
        new file:   script/HelperConfig.s.sol
        new file:   script/Interactions.s.sol
        deleted:    src/Counter.sol
        new file:   src/FundMe.sol
        new file:   src/PriceConverter.sol
        deleted:    test/Counter.t.sol
        new file:   test/Integration/InteractionsTest.t.sol
        new file:   test/mocks/MockV3Aggregator.sol
        new file:   test/unit/FundMeTest.t.sol

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   lib/chainlink-brownie-contracts (modified content)
```

Anyways, 
I proceed 
```zsh
git commit -m 'my first commit' 
```

When I run 'git commit' I saw this. 

```zsh
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   lib/chainlink-brownie-contracts (modified content)
```

Can anyone tell what is wrong with my 
```
lib/chainlink-brownie-contracts (modified content)
``` 

I think modified content means probably mock folder and files?

**Comment by 0xusmanf** [comment]
There is nothing wrong with `lib/chainlink-brownie-contracts` file. These files are just colored differently according to their status.

Run the following command before you run `git commit -m 'my first commit'`. You have to stage all the files for committing by running the following command, only then you can commit them.
```shell
git add .
```

**Reply by LareineHan** [reply]
Actually I did 
```zsh
git add.
```
but still seeing it with 
```
modified:   lib/chainlink-brownie-contracts (modified content)
```

**Reply by 0xusmanf** [reply]
Try the following command and then try to commit.

```shell
git add lib/chainlink-brownie-contracts
```

**Reply by LareineHan** [reply]
I did it and still shows like this 
```zsh
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   lib/chainlink-brownie-contracts (modified content)
```

**Reply by 0xusmanf** [reply]
Can you please show me how you are doing it step by step with screenshots? First run `git status` command then `git add .` then `git commit -m 'my first commit'` and show the results of each command with screen shots.

**Reply by LareineHan** [reply]
<img width="627" alt="Screenshot 2023-09-14 at 2 38 13 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/3efad830-efed-4676-8aaa-79cdbb425937">
This is it!

**Reply by 0xusmanf** [reply]
Have you modified anything in `lib/chainlink-brownie-contracts` folder?

I am not sure why it is giving this error, even if you modified something in this folder, git should point to that specific file instead of just pointing to the folder.

**Reply by LareineHan** [reply]
Yes, I added  'MockV3Aggregator.sol' file on 
```
lib/chainlink-brownie-contracts/contracts/src/v0.8/tests/MockV3Aggregator.sol
```
as Patrick did on the course.

**Reply by 0xusmanf** [reply]
Not sure why you are getting this error, maybe @theirrationalone, @neftyr @shikhar229169 can help you.

**Reply by theirrationalone** [reply]
Heyy @usmanfarooq91 ! Thank You for calling. 
```bash
git config pull.rebase true
git pull
```
After that, Try again with git add.

**Reply by LareineHan** [reply]
Thanks for the reply @theirrationalone and @usmanfarooq91 . However, still same problem occurs! 
This is so weird! 
```zsh
lareinehan@Lareines-MacBook-Air foundry-fund-me-f23 % git config pull.rebase true
lareinehan@Lareines-MacBook-Air foundry-fund-me-f23 % git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   lib/chainlink-brownie-contracts (modified content)

no changes added to commit (use "git add" and/or "git commit -a")
lareinehan@Lareines-MacBook-Air foundry-fund-me-f23 % git pull
Already up to date.
Current branch main is up to date.
lareinehan@Lareines-MacBook-Air foundry-fund-me-f23 % git add .
lareinehan@Lareines-MacBook-Air foundry-fund-me-f23 % git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   lib/chainlink-brownie-contracts (modified content)

no changes added to commit (use "git add" and/or "git commit -a")
```

**Reply by Niferu** [reply, marked as answer]
Hi @LareineHan 

Are you sure you are using `git add .` and not `git add.`?

Please also try to commit changes bys using below:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/8ed80957-671e-49b6-90b7-841ff2c2ca5c)

Just add some message in the box and click blue button 1 time to commit and second time to push your changes to github. Also check if there is that `lib/chainlink-brownie-contracts` file in `Changes` list.

**Reply by LareineHan** [reply]
Awesome! The bug is finally killed. Thank you @Neftyr ! 👯 
<img width="976" alt="Screenshot 2023-09-15 at 12 34 07 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/14496b7e-95b4-4e0e-9264-52aba808fc54">
<img width="482" alt="Screenshot 2023-09-15 at 12 50 16 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/128317012/1e202761-94d3-4f34-85d5-ed26cd2dc49d">

**Comment by AbdulWadood-Official** [comment]
Hi @Nlferu, Hope you're fine, I have the same error that occurred as [LareineHan], and I followed the same steps, but here I am unable to solve the terminal bugs due to these problems I have lost some important files from the git unintentionally around 125 files, 
As you can see I am on the SMIT path but the error occurred at this path: To https://github.com/Abdul-Wadood-0317/PractProject.git.
I don't know how to navigate to the current path which is:  D:\CodeforPython\PythonCodeSeries\SMIT>

Looking forward to your valuable guidance,
Thanks.
![problem1](https://github.com/Cyfrin/foundry-full-course-f23/assets/128102396/bbaa77ed-30ba-4afe-bb2f-4e53683e39a9)

**Reply by Niferu** [reply]
@AbdulWadood-Official 

Your file directory has nothing to do with your initialized git repo link. Seems like you project from path `D:\CodeforPython\PythonCodeSeries\SMIT` is just initialized and assigned to repo ` https://github.com/Abdul-Wadood-0317/PractProject.git`. If you want to change repo link just delete `.git` file from your project folder and check if `git remote -v` is empty. If it is you can just reinitialize folder with proper git repo link.

**Reply by Niferu** [reply]
The error itself explained you what the problem is. Your current local fodler has missing updates regarding to linked repository, so carefully consider if you want to match repo version code or local code should be indeed pushed into your repo.

If you are sure you want to push your local directory to assigned repo just use `git push --force origin master` here just pick whatever you are using `main/master` etc

**Reply by Moctel** [reply]
how we deal with this @Nlferu

**Comment by AbdulWadood-Official** [comment]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/128102396/7aa6e5f1-69e3-4a1f-b072-1fb0df51adf9)

@usmanfarooq91 @theirrationalone 
I am getting this error again and again when I am trying to push a commit or push code to a remote repository,
Please assist me in this regard.
Thanks

**Reply by Niferu** [reply]
Before commit you need to stage all changes (you did not staged any changes so there is nothing to commit). To do this just just click each file you want to commit like below:

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/bef37b9b-0fcc-45dd-a4ce-85dfae170ec7)

after that just add msg and click commit changes
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/1fef5cf8-6913-45db-8132-99ef6c6b072a)

you can do all of that from console level just remember to use right order

**Reply by AbdulWadood-Official** [reply]
@Nlferu Thank you so much for your detailed guide, Yes I deeply analyzed the bugs that what the console is trying to say and followed every step that you mentioned, The bug is finally killed and it is tracked with the main branch. Once again, Thanks everyone!
